src/js_native_api_v8.cc

```diff
@@ -654,72 +606,25 @@ void Reference::Finalize(bool is_env_teardown) {
 // the secondPassParameter so that even if it has been
 // scheduled no Finalization will be run.
 void Reference::ClearWeak() {
-  if (!_persistent.IsEmpty()) {
-    _persistent.ClearWeak();
-  }
-  if (_secondPassParameter != nullptr) {
-    *_secondPassParameter = nullptr;
+  if (!persistent_.IsEmpty()) {
+    persistent_.ClearWeak();
   }
 }
 
 // Mark the reference as weak and eligible for collection
 // by the gc.
 void Reference::SetWeak() {
-  if (_secondPassParameter == nullptr) {
-    // This means that the Reference has already been processed
-    // by the second pass callback, so its already been Finalized, do
-    // nothing
-    return;
-  }
-  _persistent.SetWeak(
-      _secondPassParameter, FinalizeCallback, v8::WeakCallbackType::kParameter);
-  *_secondPassParameter = this;
+  persistent_.SetWeak(this, WeakCallback, v8::WeakCallbackType::kParameter);
```

Comment:
Should we check that `persistent_` is not empty here?

Reply:
Yeah, right. It is possible that the persistent handle is cleared in the weak callback and the ref count is increased or decreased before the second pass callback. Updated.